File: go/test/project_route_entity_test.go
package sdktest

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/vercel-sdk/go"
	"github.com/voxgig-sdk/vercel-sdk/go/core"

	vs "github.com/voxgig-sdk/vercel-sdk/go/utility/struct"
)

func TestProjectRouteEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.ProjectRoute(nil)
		if ent == nil {
			t.Fatal("expected non-nil ProjectRouteEntity")
		}
	})

	// Feature #4: the entity Stream(action, ...) method runs the op pipeline and
	// returns a channel over result items. With the streaming feature active it
	// yields the feature's incremental output; otherwise it falls back to the
	// materialised list so Stream always yields.
	t.Run("stream", func(t *testing.T) {
		seed := map[string]any{
			"entity": map[string]any{
				"project_route": map[string]any{
					"s1": map[string]any{"id": "s1"},
					"s2": map[string]any{"id": "s2"},
					"s3": map[string]any{"id": "s3"},
				},
			},
		}

		// Fallback: streaming inactive -> yields the materialised list items.
		base := sdk.TestSDK(seed, nil)
		var seen []any
		for item := range base.ProjectRoute(nil).Stream("list", nil, nil) {
			seen = append(seen, item)
		}
		if len(seen) != 3 {
			t.Fatalf("expected 3 streamed items, got %d", len(seen))
		}

		// Inbound: streaming active -> yields each item from the feature iterator.
		hasStreaming := false
		if fm, ok := core.SharedConfig()["feature"].(map[string]any); ok {
			_, hasStreaming = fm["streaming"]
		}
		if hasStreaming {
			streamSdk := sdk.TestSDK(seed, map[string]any{
				"feature": map[string]any{"streaming": map[string]any{"active": true}},
			})
			var got []any
			for item := range streamSdk.ProjectRoute(nil).Stream("list", nil, nil) {
				if sub, ok := item.([]any); ok {
					got = append(got, sub...)
				} else {
					got = append(got, item)
				}
			}
			if len(got) != 3 {
				t.Fatalf("expected 3 items via streaming feature, got %d", len(got))
			}
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := project_routeBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"create", "list", "update", "load", "remove"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "project_route." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_PROJECT_ROUTE_ENTID JSON to run live")
			return
		}
		client := setup.client

		// CREATE
		projectRouteRef01Ent := client.ProjectRoute(nil)
		projectRouteRef01Data := core.ToMapAny(vs.GetProp(
			vs.GetPath(setup.data, []any{"new", "project_route"}), "project_route_ref01"))
		projectRouteRef01Data["project_id"] = setup.idmap["project01"]

		projectRouteRef01DataResult, err := projectRouteRef01Ent.Create(projectRouteRef01Data, nil)
		if err != nil {
			t.Fatalf("create failed: %v", err)
		}
		projectRouteRef01Data = core.ToMapAny(entityData(projectRouteRef01DataResult))
		if projectRouteRef01Data == nil {
			t.Fatal("expected create result to be a map")
		}
		if projectRouteRef01Data["id"] == nil {
			t.Fatal("expected created entity to have an id")
		}

		// LIST
		projectRouteRef01Match := map[string]any{
			"project_id": setup.idmap["project01"],
		}

		projectRouteRef01ListResult, err := projectRouteRef01Ent.List(projectRouteRef01Match, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		projectRouteRef01List, projectRouteRef01ListOk := projectRouteRef01ListResult.([]any)
		if !projectRouteRef01ListOk {
			t.Fatalf("expected list result to be an array, got %T", projectRouteRef01ListResult)
		}

		foundItem := vs.Select(entityListToData(projectRouteRef01List), map[string]any{"id": projectRouteRef01Data["id"]})
		if vs.IsEmpty(foundItem) {
			t.Fatal("expected to find created entity in list")
		}

		// UPDATE
		projectRouteRef01DataUp0Up := map[string]any{
			"id": projectRouteRef01Data["id"],
		}

		projectRouteRef01MarkdefUp0Name := "action"
		projectRouteRef01MarkdefUp0Value := fmt.Sprintf("Mark01-project_route_ref01_%d", setup.now)
		projectRouteRef01DataUp0Up[projectRouteRef01MarkdefUp0Name] = projectRouteRef01MarkdefUp0Value

		projectRouteRef01ResdataUp0Result, err := projectRouteRef01Ent.Update(projectRouteRef01DataUp0Up, nil)
		if err != nil {
			t.Fatalf("update failed: %v", err)
		}
		projectRouteRef01ResdataUp0 := core.ToMapAny(entityData(projectRouteRef01ResdataUp0Result))
		if projectRouteRef01ResdataUp0 == nil {
			t.Fatal("expected update result to be a map")
		}
		if projectRouteRef01ResdataUp0["id"] != projectRouteRef01DataUp0Up["id"] {
			t.Fatal("expected update result id to match")
		}
		if projectRouteRef01ResdataUp0[projectRouteRef01MarkdefUp0Name] != projectRouteRef01MarkdefUp0Value {
			t.Fatalf("expected %s to be updated, got %v", projectRouteRef01MarkdefUp0Name, projectRouteRef01ResdataUp0[projectRouteRef01MarkdefUp0Name])
		}

		// LOAD
		projectRouteRef01MatchDt0 := map[string]any{
			"id": projectRouteRef01Data["id"],
		}
		projectRouteRef01DataDt0Loaded, err := projectRouteRef01Ent.Load(projectRouteRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		projectRouteRef01DataDt0LoadResult := core.ToMapAny(entityData(projectRouteRef01DataDt0Loaded))
		if projectRouteRef01DataDt0LoadResult == nil {
			t.Fatal("expected load result to be a map")
		}
		if projectRouteRef01DataDt0LoadResult["id"] != projectRouteRef01Data["id"] {
			t.Fatal("expected load result id to match")
		}

		// REMOVE
		projectRouteRef01MatchRm0 := map[string]any{
			"id": projectRouteRef01Data["id"],
		}
		_, err = projectRouteRef01Ent.Remove(projectRouteRef01MatchRm0, nil)
		if err != nil {
			t.Fatalf("remove failed: %v", err)
		}

		// LIST
		projectRouteRef01MatchRt0 := map[string]any{
			"project_id": setup.idmap["project01"],
		}

		projectRouteRef01ListRt0Result, err := projectRouteRef01Ent.List(projectRouteRef01MatchRt0, nil)
		if err != nil {
			t.Fatalf("list failed: %v", err)
		}
		projectRouteRef01ListRt0, projectRouteRef01ListRt0Ok := projectRouteRef01ListRt0Result.([]any)
		if !projectRouteRef01ListRt0Ok {
			t.Fatalf("expected list result to be an array, got %T", projectRouteRef01ListRt0Result)
		}

		notFoundItem := vs.Select(entityListToData(projectRouteRef01ListRt0), map[string]any{"id": projectRouteRef01Data["id"]})
		if !vs.IsEmpty(notFoundItem) {
			t.Fatal("expected removed entity to not be in list")
		}

	})
}

func project_routeBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "project_route", "ProjectRouteTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read project_route test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse project_route test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap, _ := vs.Transform(
		[]any{"project_route01", "project_route02", "project_route03", "project01", "project02", "project03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("VERCEL_TEST_PROJECT_ROUTE_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"VERCEL_TEST_PROJECT_ROUTE_ENTID": idmap,
		"VERCEL_TEST_LIVE":      "FALSE",
		"VERCEL_TEST_EXPLAIN":   "FALSE",
		"VERCEL_APIKEY":         "",
	})

	idmapResolved := core.ToMapAny(env["VERCEL_TEST_PROJECT_ROUTE_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["VERCEL_TEST_LIVE"] == "TRUE" {
		// An empty map, not a nil one: Merge returns nil when its last entry
		// is nil, and BasicSetup is normally called with no extras - so a
		// bare nil silently discarded the apikey and server values below.
		extraOpts := extra
		if extraOpts == nil {
			extraOpts = map[string]any{}
		}

		mergedOpts := vs.Merge([]any{
			// liveClientOptions() FIRST, so the generated fields below win:
			// sdk-test-control.json's test.client.options adds to the live
			// client, it does not redirect it.
			liveClientOptions(),
			map[string]any{
				"apikey": env["VERCEL_APIKEY"],
			},
			extraOpts,
		})
		client = sdk.NewVercelSDK(core.ToMapAny(mergedOpts))
	}

	live := env["VERCEL_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["VERCEL_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}

File: lua/test/project_route_entity_test.lua
-- ProjectRoute entity test

local json = require("dkjson")
local vs = require("utility.struct.struct")
local sdk = require("vercel_sdk")
local helpers = require("core.helpers")
local runner = require("test.runner")

local _test_dir = debug.getinfo(1, "S").source:match("^@(.+/)")  or "./"

describe("ProjectRouteEntity", function()
  it("should create instance", function()
    local testsdk = sdk.test(nil, nil)
    local ent = testsdk:ProjectRoute(nil)
    assert.is_not_nil(ent)
  end)

  -- Feature #4: the entity stream(action, ...) method runs the op pipeline and
  -- returns an iterator over result items. With the streaming feature active it
  -- yields the feature's incremental output; otherwise it falls back to the
  -- materialised list so stream always yields.
  it("should stream", function()
    local seed = {
      entity = {
        ["project_route"] = {
          s1 = { id = "s1" },
          s2 = { id = "s2" },
          s3 = { id = "s3" },
        },
      },
    }

    -- Fallback: streaming inactive -> yields the materialised list items.
    local base = sdk.test(seed, nil)
    local seen = {}
    for item in base:ProjectRoute(nil):stream("list", nil, nil) do
      table.insert(seen, item)
    end
    assert.are.equal(3, #seen)

    -- Inbound: streaming active -> yields each item from the feature.
    local config = require("config_shared")()
    if type(config.feature) == "table" and config.feature.streaming ~= nil then
      local streamsdk = sdk.test(seed, { feature = { streaming = { active = true } } })
      local got = {}
      for item in streamsdk:ProjectRoute(nil):stream("list", nil, nil) do
        if vs.islist(item) then
          for _, sub in ipairs(item) do
            table.insert(got, sub)
          end
        else
          table.insert(got, item)
        end
      end
      assert.are.equal(3, #got)
    end
  end)

  it("should run basic flow", function()
    local setup = project_route_basic_setup(nil)
    -- Per-op sdk-test-control.json skip.
    local _live = setup.live or false
    for _, _op in ipairs({"create", "list", "update", "load", "remove"}) do
      local _should_skip, _reason = runner.is_control_skipped("entityOp", "project_route." .. _op, _live and "live" or "unit")
      if _should_skip then
        pending(_reason or "skipped via sdk-test-control.json")
        return
      end
    end
    -- The basic flow consumes synthetic IDs from the fixture. In live mode
    -- without an *_ENTID env override, those IDs hit the live API and 4xx.
    if setup.synthetic_only then
      pending("live entity test uses synthetic IDs from fixture — set VERCEL_TEST_PROJECT_ROUTE_ENTID JSON to run live")
      return
    end
    local client = setup.client

    -- CREATE
    local project_route_ref01_ent = client:ProjectRoute(nil)
    local project_route_ref01_data = helpers.to_map(vs.getprop(
      vs.getpath(setup.data, "new.project_route"), "project_route_ref01"))
    project_route_ref01_data["project_id"] = setup.idmap["project01"]

    local project_route_ref01_data_result, err = project_route_ref01_ent:create(project_route_ref01_data, nil)
    assert.is_nil(err)
    project_route_ref01_data = helpers.to_map(type(project_route_ref01_data_result) == 'table' and project_route_ref01_data_result.data_get and project_route_ref01_data_result:data_get() or project_route_ref01_data_result)
    assert.is_not_nil(project_route_ref01_data)
    assert.is_not_nil(project_route_ref01_data["id"])

    -- LIST
    local project_route_ref01_match = {
      ["project_id"] = setup.idmap["project01"],
    }

    local project_route_ref01_list_result, err = project_route_ref01_ent:list(project_route_ref01_match, nil)
    assert.is_nil(err)
    assert.is_table(project_route_ref01_list_result)

    local found_item = vs.select(
      runner.entity_list_to_data(project_route_ref01_list_result),
      { id = project_route_ref01_data["id"] })
    assert.is_false(vs.isempty(found_item))

    -- UPDATE
    local project_route_ref01_data_up0_up = {
      id = project_route_ref01_data["id"],
    }

    local project_route_ref01_markdef_up0_name = "action"
    local project_route_ref01_markdef_up0_value = "Mark01-project_route_ref01_" .. tostring(setup.now)
    project_route_ref01_data_up0_up[project_route_ref01_markdef_up0_name] = project_route_ref01_markdef_up0_value

    local project_route_ref01_resdata_up0_result, err = project_route_ref01_ent:update(project_route_ref01_data_up0_up, nil)
    assert.is_nil(err)
    local project_route_ref01_resdata_up0 = helpers.to_map(type(project_route_ref01_resdata_up0_result) == 'table' and project_route_ref01_resdata_up0_result.data_get and project_route_ref01_resdata_up0_result:data_get() or project_route_ref01_resdata_up0_result)
    assert.is_not_nil(project_route_ref01_resdata_up0)
    assert.are.equal(project_route_ref01_resdata_up0["id"], project_route_ref01_data_up0_up["id"])
    assert.are.equal(project_route_ref01_resdata_up0[project_route_ref01_markdef_up0_name], project_route_ref01_markdef_up0_value)

    -- LOAD
    local project_route_ref01_match_dt0 = {
      id = project_route_ref01_data["id"],
    }
    local project_route_ref01_data_dt0_loaded, err = project_route_ref01_ent:load(project_route_ref01_match_dt0, nil)
    assert.is_nil(err)
    local project_route_ref01_data_dt0_load_result = helpers.to_map(type(project_route_ref01_data_dt0_loaded) == 'table' and project_route_ref01_data_dt0_loaded.data_get and project_route_ref01_data_dt0_loaded:data_get() or project_route_ref01_data_dt0_loaded)
    assert.is_not_nil(project_route_ref01_data_dt0_load_result)
    assert.are.equal(project_route_ref01_data_dt0_load_result["id"], project_route_ref01_data["id"])

    -- REMOVE
    local project_route_ref01_match_rm0 = {
      id = project_route_ref01_data["id"],
    }
    local _, err = project_route_ref01_ent:remove(project_route_ref01_match_rm0, nil)
    assert.is_nil(err)

    -- LIST
    local project_route_ref01_match_rt0 = {
      ["project_id"] = setup.idmap["project01"],
    }

    local project_route_ref01_list_rt0_result, err = project_route_ref01_ent:list(project_route_ref01_match_rt0, nil)
    assert.is_nil(err)
    assert.is_table(project_route_ref01_list_rt0_result)

    local not_found_item = vs.select(
      runner.entity_list_to_data(project_route_ref01_list_rt0_result),
      { id = project_route_ref01_data["id"] })
    assert.is_true(vs.isempty(not_found_item))

  end)
end)

function project_route_basic_setup(extra)
  runner.load_env_local()

  local entity_data_file = _test_dir .. "../../.sdk/test/entity/project_route/ProjectRouteTestData.json"
  local f = io.open(entity_data_file, "r")
  if f == nil then
    error("failed to read project_route test data: " .. entity_data_file)
  end
  local entity_data_source = f:read("*a")
  f:close()

  local entity_data = json.decode(entity_data_source)

  local options = {}
  options["entity"] = entity_data["existing"]

  local client = sdk.test(options, extra)

  -- Generate idmap via transform.
  local idmap = vs.transform(
    { "project_route01", "project_route02", "project_route03", "project01", "project02", "project03" },
    {
      ["`$PACK`"] = { "", {
        ["`$KEY`"] = "`$COPY`",
        ["`$VAL`"] = { "`$FORMAT`", "upper", "`$COPY`" },
      }},
    }
  )

  -- Detect ENTID env override before envOverride consumes it. When live
  -- mode is on without a real override, the basic test runs against synthetic
  -- IDs from the fixture and 4xx's. Surface this so the test can skip.
  local entid_env_raw = os.getenv("VERCEL_TEST_PROJECT_ROUTE_ENTID")
  local idmap_overridden = entid_env_raw ~= nil and entid_env_raw:match("^%s*{") ~= nil

  local env = runner.env_override({
    ["VERCEL_TEST_PROJECT_ROUTE_ENTID"] = idmap,
    ["VERCEL_TEST_LIVE"] = "FALSE",
    ["VERCEL_TEST_EXPLAIN"] = "FALSE",
    ["VERCEL_APIKEY"] = "",
  })

  local idmap_resolved = helpers.to_map(
    env["VERCEL_TEST_PROJECT_ROUTE_ENTID"])
  if idmap_resolved == nil then
    idmap_resolved = helpers.to_map(idmap)
  end

  if env["VERCEL_TEST_LIVE"] == "TRUE" then
    local merged_opts = vs.merge({
      -- FIRST, so the generated fields below win: sdk-test-control.json's
      -- test.client.options adds to the live client, it does not redirect it.
      runner.live_client_options(),
      {
        apikey = env["VERCEL_APIKEY"],
      },
      extra or {},
    })
    client = sdk.new(helpers.to_map(merged_opts))
  end

  local live = env["VERCEL_TEST_LIVE"] == "TRUE"
  return {
    client = client,
    data = entity_data,
    idmap = idmap_resolved,
    env = env,
    explain = env["VERCEL_TEST_EXPLAIN"] == "TRUE",
    live = live,
    synthetic_only = live and not idmap_overridden,
    now = os.time() * 1000,
  }
end
